### ISSUE TITLE
DateTime should have a const "epoch" constructor

### ISSUE THREAD

**Issue by Ludonope** [dart-lang/sdk#50154]
There are a few use-cases where having a const `DateTime` constructor can be useful, such as default parameter's values.

I understand why the current implementation doesn't have const constructors (there is a need for some runtime checks), however having an `epoch` constructor would allow to use a constant "zero-value" for a `DateTime` object which would directly create the object without going through the dynamic checks.

**Comment by Anant-2005** [comment]
Hi Sir, Can you please assign me this task. I will love to work on it :-)

**Comment by lrhn** [comment]
The class itself needs to be constant for that to be possible. It looks like that in the source, but the individual platforms patch that class with more fields which aren't even final (for caching the individual getter values the first time you ask for them, [like the VM does](https://github.com/dart-lang/sdk/blob/6524b3196742a46a7a0b76031682bdddcd1ba3b3/sdk/lib/_internal/vm/lib/date_patch.dart#L32)).

The `epoch` value itself could be a different class, with all getters hardwirded to return 0, 1 or 1970 as appropriate.
That risks polymorphism, which may reduce the performance of other code. Not impossible, but less convenient.

Or there could be another class for constant uses, which doesn't cache and is therefore less efficient on repeated calls. (Computing the year usually also implies computing the month and day, which is why caching them all on the first call makes sense.)

It's just not a very good class to make constant, there are too many computations *either* at creation or when inspecting.

**Comment by Ludonope** [comment]
If the year/month/day were calculated on inspection, when passing constant values, wouldn't those basically be set at compilation?

I'm not sure of the level of optimization of the Dart compiler but if it has a math literal with all values already set it should be able to resolve it at compilation. I might be missing some pieces tho :)

**Comment by lrhn** [comment]
Computing the Y/M/D/h/m/s/S values at constant creation time is probably technically possible for UTC dates. It's definitely not possible for local time, since we don't know the local time where the code will run. It's also a complicated computation, that is hard to fit into constant constructors using only potentially constant expressions and no complicated control flow. It's also hard to introduce bounds checks in constant computations.

It still needs a new class to hold those, since the current implementation class has a mutable field, and therefore cannot be constant.

All in all, not going to be easy or generally applicable, and probably not really worth the effort.

**Comment by westy92** [comment]
You can use my [const_date_time](https://pub.dev/packages/const_date_time) package to achieve this.

```dart
import 'package:const_date_time/const_date_time.dart';

class MyClass {
  const MyClass({
    this.date = const ConstDateTime(0),
  });
  
  final DateTime date;
}
```

**Comment by ahmednfwela** [comment]
> for caching the individual getter values the first time you ask for them

@lrhn would using `Expando` be useful here ?

**Comment by lrhn** [comment]
Using `Expando` is possible, but not without cost. It's just not as efficient.

Another reason to be against having an `epoch` constant is that it's just not a very good default value.
There is nothing special about the epoch as a point time. It's an arbitrary date chosen as a base number for the implementation, but using that as a default value in an API is very likely a mistake. You will want a default value that is either clearly recognizable as not a real value (the epoch is still the birthday of real people, using 1st of January year zero would be a better "no value" value for that use case).

So, for a default value, you really should make the parameter nullable and accept `null`. Using a perfectly reasonable value as default risks either testing that value indirectly if it's a real argument, or assuming the default value means something.
In short, the epoch is not special as a date, it's an artifact of an implementation choice. There is no situation where you should care what the epoch date is, unless you want to use the underlying integer representation.

Not going to do an `epoch` constant.
Making the class constant in general is maybe possible, but not for local time, and the API, and the currently required compressions in the constructors, really isn't geared to being constant. Like `RegExp`, it does too much validation in the constructor to be a good constant type.
